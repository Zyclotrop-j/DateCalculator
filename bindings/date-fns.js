import {
  parseISO,
  isValid,
  differenceInMilliseconds,
  isDate,
  toDate,
  addMilliseconds,
  addSeconds,
  addMinutes,
  addHours,
  addBusinessDays,
  addDays,
  addWeeks,
  addMonths,
  addQuarters,
  addYears,
  subMilliseconds,
  subSeconds,
  subMinutes,
  subHours,
  subBusinessDays,
  subDays,
  subWeeks,
  subMonths,
  subQuarters,
  subYears
} from "date-fns";

const divTable = {
  milliseconds: 1,
  seconds: 1000,
  minutes: 1000 * 60,
  hours: 1000 * 60 * 60,
  days: 1000 * 60 * 60 * 24,
  weekdays: 1000 * 60 * 60 * 24,
  weeks: 1000 * 60 * 60 * 24 * 7,
  months: 1000 * 60 * 60 * 24 * 30.436875,
  quarters: 1000 * 60 * 60 * 24 * 91.310625,
  years: 1000 * 60 * 60 * 24 * 365.2425
};
Object.fromEntries =
  Object.fromEntries ||
  (arr =>
    [...arr].reduce((obj, [key, val]) => {
      obj[key] = val;
      return obj;
    }, {}));

const ops = Object.fromEntries(
  Object.entries({
    addMilliseconds,
    addSeconds,
    addMinutes,
    addHours,
    addWeekDays: addBusinessDays,
    addDays,
    addWeeks,
    addMonths,
    addQuarters,
    addYears,
    subMilliseconds,
    subSeconds,
    subMinutes,
    subHours,
    subWeekDays: subBusinessDays,
    subDays,
    subWeeks,
    subMonths,
    subQuarters,
    subYears,

    addUnitless: addMilliseconds,
    subUnitless: subMilliseconds
  }).map(([k, v]) => [k.toLowerCase(), v])
);

export const datefns = (
  { NATIVEDATE, DATEXPRESION, DURATIONEXPRESSION, DURATIONOBJECT },
  { console }
) => {
  const tostring = {
    configurable: false,
    enumerable: false,
    writable: false,
    value: function() {
      return JSON.stringify(this);
    }
  };
  const isDurationConfig = {
    configurable: false,
    enumerable: false,
    writable: false,
    value: true
  };
  const isDurationSymbol = Symbol("isDuration");
  const newDuration = () =>
    Object.defineProperty(
      Object.defineProperty({}, isDurationSymbol, isDurationConfig),
      "toString",
      tostring
    );
  const isDuration = x => x[isDurationSymbol] === true;
  function makeDuration(arg) {
    const isoRegex = /^(-|\+)?P(?:([-+]?[0-9,.]*)Y)?(?:([-+]?[0-9,.]*)M)?(?:([-+]?[0-9,.]*)W)?(?:([-+]?[0-9,.]*)D)?(?:T(?:([-+]?[0-9,.]*)H)?(?:([-+]?[0-9,.]*)M)?(?:([-+]?[0-9,.]*)S)?)?$/;
    const match = isoRegex.exec(arg);
    if (match) {
      const sign = match[1] === "-" ? -1 : 1;
      const parseFloatInIso = mat => {
        const res = mat && parseFloat(mat.replace(",", "."));
        return (isNaN(res) ? 0 : res) * sign;
      };
      return Object.assign(newDuration(), {
        years: parseFloatInIso(match[2]),
        months: parseFloatInIso(match[3]),
        weeks: parseFloatInIso(match[4]),
        days: parseFloatInIso(match[5]),
        hours: parseFloatInIso(match[6]),
        minutes: parseFloatInIso(match[7]),
        seconds: parseFloatInIso(match[8])
      });
    }
    try {
      return Object.assign(newDuration(), JSON.parse(arg));
    } catch (e) {
      throw new Error("Found invalid duration " + arg);
    }
  }

  const getType = arg => {
    if (isValid(arg)) return "DATE";
    if (isDuration(arg)) return "DURATION";
    if (isDate(toDate(arg)) && !isValid(arg)) return "INVALID DATE";
    return typeof arg;
  };
  const isUnitless = a => {
    const keys = Object.keys(a);
    return keys.length === 1 && keys[0] === "unitless";
  };

  return {
    makeDuration: (duration, { type }) => {
      return {
        [DURATIONEXPRESSION]: makeDuration,
        [DURATIONOBJECT]: a => Object.assign(newDuration(), a)
      }[type](duration);
    },
    makeDate: (date, { type }) => {
      return {
        [DATEXPRESION]: parseISO,
        [NATIVEDATE]: d => d
      }[type](date);
    },
    add: (a, b) => {
      if (isUnitless(a) && isUnitless(b)) {
        return { unitless: a.unitless + b.unitless };
      }
      if (isDuration(a) && isDuration(b)) {
        return Object.entries(a).reduce(
          (p, [k, v]) =>
            Object.assign(p, {
              [k]: (p[k] || 0) + v
            }),
          b
        );
      }
      if (isValid(a) && isValid(b)) {
        throw new Error("Can't add dates to each other!");
      }
      if (isValid(a) && isDuration(b)) {
        // b is duratoin
        return Object.entries(b).reduce(
          (date, [k, v]) => ops[`add${k}`](date, v),
          a
        );
      }
      if (isValid(b) && isDuration(a)) {
        // a is duratoin
        return Object.entries(a).reduce(
          (date, [k, v]) => ops[`add${k}`](date, v),
          b
        );
      }
      const e = `Invalid arguments for 'add', expected (date, duration), (duration, date), (duration, duration) but found (${getType(
        a
      )}, ${getType(b)})`;
      console.error(e, a, b);
      throw new Error(e);
    },
    subtract: (a, b) => {
      if (isUnitless(a) && isUnitless(b)) {
        return { unitless: a.unitless - b.unitless };
      }
      if (isDuration(a) && isDuration(b)) {
        return Object.entries(a).reduce(
          (p, [k, v]) =>
            Object.assign(p, {
              [k]: (p[k] || 0) - v
            }),
          b
        );
      }
      if (isValid(a) && isValid(b)) {
        console.warn(
          "subtracting two dates from each other is potentially unsafe as a loss of information occurs!"
        );
        return Object.assign(newDuration(), {
          milliseconds: differenceInMilliseconds(a, b)
        });
      }
      if (isValid(a) && isDuration(b)) {
        // b is duratoin
        return Object.entries(b).reduce(
          (date, [k, v]) => ops[`sub${k}`](date, v),
          a
        );
      }
      if (isValid(b) && isDuration(a)) {
        // a is duratoin
        return Object.entries(a).reduce(
          (date, [k, v]) => ops[`sub${k}`](date, v),
          b
        );
      }
      const e = `Invalid arguments for 'subtract', expected (date, duration), (duration, date), (duration, duration) but found (${getType(
        a
      )}, ${getType(b)})`;
      console.error(e, a, b);
      throw new Error(e);
    },
    multiply: (a, b) => {
      if (isUnitless(a) && isUnitless(b)) {
        return { unitless: a.unitless * b.unitless };
      }
      if (isDuration(a) && isUnitless(b)) {
        return Object.entries(a).reduce(
          (p, [k, v]) =>
            Object.assign(p, {
              [k]: v * b.unitless
            }),
          newDuration()
        );
      }
      if (isDuration(b) && isUnitless(a)) {
        return Object.entries(b).reduce(
          (p, [k, v]) =>
            Object.assign(p, {
              [k]: v * a.unitless
            }),
          newDuration()
        );
      }
      const e = `Invalid arguments for multiply, expected (duration, unitless) or (unitless, duration) but found (${getType(
        a
      )}, ${getType(b)})`;
      console.error(e, a, b);
      throw new Error(e);
    },
    divide: (a, b) => {
      if (isUnitless(a) && isUnitless(b)) {
        return { unitless: a.unitless / b.unitless };
      }
      if (isDuration(a) && isUnitless(b)) {
        return Object.entries(a).reduce(
          (p, [k, v]) =>
            Object.assign(p, {
              [k]: v / b.unitless
            }),
          newDuration()
        );
      }
      if (isDuration(b) && isUnitless(a)) {
        return Object.entries(b).reduce(
          (p, [k, v]) =>
            Object.assign(p, {
              [k]: a.unitless / v
            }),
          newDuration()
        );
      }
      if (isDuration(a) && isDuration(b)) {
        console.warn(
          `Dividing one duration by another is potentially unsafe! Assuming conversion table ${JSON.stringify(
            divTable,
            null,
            "  "
          )}!`
        );
        const bInMs = Object.entries(b).reduce(
          (p, [k, v]) => p + divTable[k] * v,
          0
        );
        const aInMs = Object.entries(a).reduce(
          (p, [k, v]) => p + divTable[k] * v,
          0
        );
        return { unitless: aInMs / bInMs };
      }
      const e = `Invalid arguments for divide, expected (duration, unitless) or (unitless, duration) but found (${getType(
        a
      )}, ${getType(b)})`;
      console.error(e, a, b);
      throw new Error(e);
    }
  };
};
export default datefns;
