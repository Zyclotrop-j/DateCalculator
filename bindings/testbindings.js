export const test = (
  __,
  { makeDuration, makeDate, add, substract, multiply, divide }
) => {
  return {
    makeDuration: makeDuration,
    makeDate: makeDate,
    add: add,
    substract: substract,
    multiply: multiply,
    divide: divide
  };
};
export default test;
