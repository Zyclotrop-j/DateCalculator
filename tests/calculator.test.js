import parser from "../dateCalculator";
import { test as testbinding } from "../bindings/testbindings";

const it = test;

describe("Timezones", () => {
  it("should always be UTC", () => {
    expect(new Date().getTimezoneOffset()).toBe(0);
  });
});

it("Instanciation works", () => {
  expect(testbinding({}, {})).toEqual({});
  expect(parser(testbinding(parser, {}))).toBeInstanceOf(Function);
});

describe("Parsing without a binding", () => {
  const RealDate = Date;
  let calculator;
  const operations = {};

  beforeAll(() => {
    const DATE_TO_USE = new Date();
    global.Date = jest.fn((str, ...args) => {
      if (!str) return DATE_TO_USE;
      return new RealDate(str, ...args);
    });
    global.Date.UTC = RealDate.UTC;
    global.Date.parse = RealDate.parse;
    global.Date.now = RealDate.now;
  });
  beforeEach(() => {
    operations.makeDuration = jest.fn(duration => ({ duration }));
    operations.makeDate = jest.fn(date => ({ date }));
    operations.add = jest.fn((a, b) => ({ add: [a, b] }));
    operations.substract = jest.fn((a, b) => ({ substract: [a, b] }));
    operations.multiply = jest.fn((a, b) => ({ multiply: [a, b] }));
    operations.divide = jest.fn((a, b) => ({ divide: [a, b] }));
    calculator = parser(testbinding(parser, operations));
  });
  afterEach(() => {
    calculator = null;
  });
  afterAll(() => {
    global.Date = RealDate;
  });

  it(" returns new date for 'now'", () => {
    expect(calculator`now`).toEqual({ date: new Date() });
    expect(calculator("now")).toEqual({ date: new Date() });
  });

  it(" returns string-interpolations literally", () => {
    expect(calculator`${"now"}`).toEqual({ date: new Date() });
  });

  it(" returns other interpolations as-is", () => {
    expect(calculator`${["Hello World"]}`).toEqual(["Hello World"]);
  });

  it(" throw when invalid input is supplied", () => {
    expect(() => calculator``).toThrow();
    expect(() => calculator`Hello World`).toThrow();
    expect(() => calculator()).toThrow();
    expect(() => calculator("Hello World")).toThrow();
  });

  it(" parses unitless numbers", () => {
    expect(calculator`1`).toEqual({ unitless: 1 });
    expect(calculator`1u`).toEqual({ unitless: 1 });
  });

  const durationIdentifiers = {
    millisecond: ["ms"],
    second: ["s", "S", "ss", "SS", "sec", "SEC"],
    minute: ["min", "tm", "mm", "MIN", "TM"],
    hour: ["h", "H"],
    day: ["d", "D"],
    weekday: ["wd", "WD"],
    week: ["w", "W"],
    month: ["M", "MM", "l", "L"],
    quarter: ["q", "Q"],
    year: ["y", "a", "Y", "A"]
  };

  describe("Singular long-form duration identifiers", () => {
    Object.keys(durationIdentifiers).forEach(key => {
      ["", " ", "\n", "\r", "\t", " \n\r\t"].forEach(whitespace => {
        const whitespaceDescription = whitespace
          .split("")
          .map(
            i =>
              ({
                " ": "_",
                "\t": "\\t",
                "\n": "\\n",
                "\r": "\\r"
              }[i])
          )
          .join("");
        it(` parses "${key}" with whitespace "${whitespaceDescription}" correctly`, () => {
          expect(calculator`1${whitespace}${key}`).toEqual({
            duration: {
              [`${key}s`]: 1
            }
          });
          expect(calculator`0${whitespace}${key}`).toEqual({
            duration: {
              [`${key}s`]: 0
            }
          });
          expect(calculator`2${whitespace}${key}`).toEqual({
            duration: {
              [`${key}s`]: 2
            }
          });
          expect(calculator`10000${whitespace}${key}`).toEqual({
            duration: {
              [`${key}s`]: 10000
            }
          });

          expect(() => calculator`-1${whitespace}${key}`).toThrow();
          expect(() => calculator`n${whitespace}${key}`).toThrow();
        });
      });
    });
  });

  describe("Plural long-form duration identifiers", () => {
    Object.keys(durationIdentifiers).forEach(key => {
      ["", " ", "  ", "\t", "\t\t", "\t \t ", "\n", " \n \t"].forEach(
        whitespace => {
          const whitespaceDescription = whitespace
            .split("")
            .map(
              i =>
                ({
                  " ": "_",
                  "\t": "\\t",
                  "\n": "\\n"
                }[i])
            )
            .join("");
          it(` parses "${key}" with whitespace "${whitespaceDescription}" correctly`, () => {
            expect(calculator`1${whitespace}${key}s`).toEqual({
              duration: {
                [`${key}s`]: 1
              }
            });
            expect(calculator`0${whitespace}${key}s`).toEqual({
              duration: {
                [`${key}s`]: 0
              }
            });
            expect(calculator`2${whitespace}${key}s`).toEqual({
              duration: {
                [`${key}s`]: 2
              }
            });
            expect(calculator`10000${whitespace}${key}s`).toEqual({
              duration: {
                [`${key}s`]: 10000
              }
            });

            expect(() => calculator`-1${whitespace}${key}s`).toThrow();
            expect(() => calculator`n${whitespace}${key}s`).toThrow();
          });
        }
      );
    });
  });

  describe("Singular short-form duration identifiers", () => {
    Object.entries(durationIdentifiers).forEach(([ident, keys]) => {
      ["", " ", "  ", "\t", "\t\t", "\t \t ", "\n", " \n \t"].forEach(
        whitespace => {
          const whitespaceDescription = whitespace
            .split("")
            .map(
              i =>
                ({
                  " ": "_",
                  "\t": "\\t",
                  "\n": "\\n"
                }[i])
            )
            .join("");
          keys.forEach(key => {
            it(` parses "${key}" (${ident}) with whitespace "${whitespaceDescription}" correctly`, () => {
              expect(calculator`1${whitespace}${key}`).toEqual({
                duration: {
                  [`${ident}s`]: 1
                }
              });
              expect(calculator`0${whitespace}${key}`).toEqual({
                duration: {
                  [`${ident}s`]: 0
                }
              });
              expect(calculator`2${whitespace}${key}`).toEqual({
                duration: {
                  [`${ident}s`]: 2
                }
              });
              expect(calculator`10000${whitespace}${key}`).toEqual({
                duration: {
                  [`${ident}s`]: 10000
                }
              });

              expect(() => calculator`-1${whitespace}${key}`).toThrow();
              expect(() => calculator`n${whitespace}${key}`).toThrow();
            });
          });
        }
      );
    });
  });

  describe("Recognizes '+'", () => {
    const cop = "add";
    it(" calls binding callbacks for two operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1d + 1d`;
      expect(mockCallback.mock.calls.length).toBe(1);
      expect(duration.mock.calls.length).toBe(2);
      expect(duration.mock.calls[0][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[0][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(duration.mock.calls[1][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[1][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(result).toEqual({
        [cop]: [duration.mock.results[0].value, duration.mock.results[1].value]
      });
    });
    it(" calls binding callbacks for many operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1min + 1s + 1d + 1h`;
      expect(mockCallback.mock.calls.length).toBe(3);
      expect(duration.mock.calls.length).toBe(4);
      // Left-to-right evaluation
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(mockCallback.mock.calls[1][0]).toBe(
        mockCallback.mock.results[0].value
      );
      expect(mockCallback.mock.calls[1][1]).toBe(
        duration.mock.results[2].value
      );
      expect(mockCallback.mock.calls[2][0]).toBe(
        mockCallback.mock.results[1].value
      );
      expect(mockCallback.mock.calls[2][1]).toBe(
        duration.mock.results[3].value
      );
      expect(result).toEqual({
        [cop]: [
          mockCallback.mock.results[1].value,
          duration.mock.results[3].value
        ]
      });
    });
  });

  describe("Recognizes '-'", () => {
    const cop = "substract";
    it(" calls binding callbacks for two operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1d - 1d`;
      expect(mockCallback.mock.calls.length).toBe(1);
      expect(duration.mock.calls.length).toBe(2);
      expect(duration.mock.calls[0][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[0][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(duration.mock.calls[1][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[1][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(result).toEqual({
        [cop]: [duration.mock.results[0].value, duration.mock.results[1].value]
      });
    });
    it(" calls binding callbacks for many operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1min - 1s - 1d - 1h`;
      expect(mockCallback.mock.calls.length).toBe(3);
      expect(duration.mock.calls.length).toBe(4);
      // Left-to-right evaluation
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(mockCallback.mock.calls[1][0]).toBe(
        mockCallback.mock.results[0].value
      );
      expect(mockCallback.mock.calls[1][1]).toBe(
        duration.mock.results[2].value
      );
      expect(mockCallback.mock.calls[2][0]).toBe(
        mockCallback.mock.results[1].value
      );
      expect(mockCallback.mock.calls[2][1]).toBe(
        duration.mock.results[3].value
      );
      expect(result).toEqual({
        [cop]: [
          mockCallback.mock.results[1].value,
          duration.mock.results[3].value
        ]
      });
    });
  });

  describe("Recognizes '*'", () => {
    const cop = "multiply";
    it(" calls binding callbacks for two operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1d * 1d`;
      expect(mockCallback.mock.calls.length).toBe(1);
      expect(duration.mock.calls.length).toBe(2);
      expect(duration.mock.calls[0][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[0][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(duration.mock.calls[1][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[1][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(result).toEqual({
        [cop]: [duration.mock.results[0].value, duration.mock.results[1].value]
      });
    });
    it(" calls binding callbacks for many operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1min * 1s * 1d * 1h`;
      expect(mockCallback.mock.calls.length).toBe(3);
      expect(duration.mock.calls.length).toBe(4);
      // Left-to-right evaluation
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(mockCallback.mock.calls[1][0]).toBe(
        mockCallback.mock.results[0].value
      );
      expect(mockCallback.mock.calls[1][1]).toBe(
        duration.mock.results[2].value
      );
      expect(mockCallback.mock.calls[2][0]).toBe(
        mockCallback.mock.results[1].value
      );
      expect(mockCallback.mock.calls[2][1]).toBe(
        duration.mock.results[3].value
      );
      expect(result).toEqual({
        [cop]: [
          mockCallback.mock.results[1].value,
          duration.mock.results[3].value
        ]
      });
    });
  });

  describe("Recognizes '/'", () => {
    const cop = "divide";
    it(" calls binding callbacks for two operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1d / 1d`;
      expect(mockCallback.mock.calls.length).toBe(1);
      expect(duration.mock.calls.length).toBe(2);
      expect(duration.mock.calls[0][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[0][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(duration.mock.calls[1][0]).toEqual({ days: 1 });
      expect(duration.mock.calls[1][1]).toEqual({
        type: parser.DURATIONOBJECT
      });
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(result).toEqual({
        [cop]: [duration.mock.results[0].value, duration.mock.results[1].value]
      });
    });
    it(" calls binding callbacks for many operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1min / 1s / 1d / 1h`;
      expect(mockCallback.mock.calls.length).toBe(3);
      expect(duration.mock.calls.length).toBe(4);
      // Left-to-right evaluation
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[1].value
      );
      expect(mockCallback.mock.calls[1][0]).toBe(
        mockCallback.mock.results[0].value
      );
      expect(mockCallback.mock.calls[1][1]).toBe(
        duration.mock.results[2].value
      );
      expect(mockCallback.mock.calls[2][0]).toBe(
        mockCallback.mock.results[1].value
      );
      expect(mockCallback.mock.calls[2][1]).toBe(
        duration.mock.results[3].value
      );
      expect(result).toEqual({
        [cop]: [
          mockCallback.mock.results[1].value,
          duration.mock.results[3].value
        ]
      });
    });
  });

  describe("Recognizes '()'", () => {
    const cop = "add";
    it(" calls binding callbacks for two operands", () => {
      const mockCallback = operations[cop];
      const duration = operations.makeDuration;
      const result = calculator`1s + (1min + 1d)`;
      expect(mockCallback.mock.calls.length).toBe(2);
      expect(duration.mock.calls.length).toBe(3);
      // right first, then left-to-right
      expect(mockCallback.mock.calls[0][0]).toBe(
        duration.mock.results[1].value
      );
      expect(mockCallback.mock.calls[0][1]).toBe(
        duration.mock.results[2].value
      );
      expect(mockCallback.mock.calls[1][0]).toBe(
        duration.mock.results[0].value
      );
      expect(mockCallback.mock.calls[1][1]).toBe(
        mockCallback.mock.results[0].value
      );
      expect(result).toEqual({
        add: [
          duration.mock.results[0].value,
          mockCallback.mock.results[0].value
        ]
      });
    });
    it(" applys '*' and '/' before '+' and '-'", () => {
      const result1 = calculator`1s + 1min * 1d`;
      expect(result1).toEqual({
        add: [
          operations.makeDuration.mock.results[0].value,
          operations.multiply.mock.results[0].value
        ]
      });
      const result2 = calculator`1s - 1min / 1d`;
      expect(result2).toEqual({
        substract: [
          operations.makeDuration.mock.results[0].value,
          operations.divide.mock.results[0].value
        ]
      });
    });
    it(" applys '()' first", () => {
      const result1 = calculator`(1s + 1min) * 1d`;
      expect(result1).toEqual({
        multiply: [
          operations.add.mock.results[0].value,
          operations.makeDuration.mock.results[2].value
        ]
      });
      const result2 = calculator`(1s - 1min) / 1d`;
      expect(result2).toEqual({
        divide: [
          operations.substract.mock.results[0].value,
          operations.makeDuration.mock.results[2].value
        ]
      });
    });
  });

  describe("Parses iso-dates", () => {
    const extended = [
      "2009-12T12:34",
      "2009",
      "2009-05-19",
      " 2009-05-19 ",
      "2009-05",
      "2009-001",
      "2009-05-19",
      "2009-05-19T00:00",
      "2009-05-19T14:31",
      "2009-05-19T14:39:22",
      "2009-05-19T14:39Z",
      "2009-05-19T14:39:22-06:00",
      "2009-05-19T14:39:22+0600",
      "2007-04-06T00:00"
    ];
    const iso = [
      "200912T1234",
      "2009",
      "20090519",
      " 20090519 ",
      "200905",
      "2009001",
      "20090519",
      "20090519 0000",
      "20090519 1431",
      "20090519 143922",
      "20090519 1439Z",
      "20090519 143922-06:00",
      "20090519 143922+0600",
      "20070406 0000"
    ];
    extended.forEach(i =>
      it(` parse "${i}" successfully`, () => {
        const result = calculator`${i}`;
        expect(result).toBeDefined();
        expect(`${i} = ${result.date}`).toMatchSnapshot();
      })
    );
    iso.forEach((i, idx) =>
      it(` parse "${i}" successfully`, () => {
        const result = calculator`${i}`;
        expect(result).toBeDefined();
        expect(`${i} = ${result.date}`).toMatchSnapshot();
      })
    );

    const invalidisos = [
      "200913",
      "2009367",
      "2009-",
      "20070405T24:50",
      // "2009-000", // substract operation!! -> valid input
      "2009-M511",
      "2009M511",
      "2009-05-19T14a39r",
      "2009-05-1914:39",
      "2009-05-19 14:",
      "2009-05-19r14:39",
      "2009-05-19 14a39a22",
      "2009-05-19 14:39:22+06a00",
      "2009-05-19 146922.500",
      "2010-02-18T16.5:23.35:48",
      "2010-02-18T16:23.35:48",
      "2010-02-18T16:23.35:48.45",
      "2009-05-19 14.5.44",
      "2010-02-18T16:23.33.600",
      "2010-02-18T16,25:23:48,444"
    ];
    invalidisos.forEach((i, idx) =>
      it(` throws on "${i}"`, () => {
        const result = () => calculator`${i}`;
        expect(result).toThrow();
      })
    );
  });

  describe("Parses iso-durations", () => {});

  describe("Parses iso-intervals", () => {});
});
