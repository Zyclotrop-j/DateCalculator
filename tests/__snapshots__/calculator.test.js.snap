// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parsing without a binding Parses iso-dates  parse " 2009-05-19 " successfully 1`] = `" 2009-05-19  = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse " 20090519 " successfully 1`] = `" 20090519  = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2007-04-06T00:00" successfully 1`] = `"2007-04-06T00:00 = Fri Apr 06 2007 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009" successfully 1`] = `"2009 = Thu Jan 01 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009" successfully 2`] = `"2009 = Thu Jan 01 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-001" successfully 1`] = `"2009-001 = Fri Jan 02 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05" successfully 1`] = `"2009-05 = Fri May 01 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19" successfully 1`] = `"2009-05-19 = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19" successfully 2`] = `"2009-05-19 = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19T00:00" successfully 1`] = `"2009-05-19T00:00 = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19T14:31" successfully 1`] = `"2009-05-19T14:31 = Tue May 19 2009 14:31:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19T14:39:22" successfully 1`] = `"2009-05-19T14:39:22 = Tue May 19 2009 14:39:22 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19T14:39:22+0600" successfully 1`] = `"2009-05-19T14:39:22+0600 = Tue May 19 2009 08:39:22 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19T14:39:22-06:00" successfully 1`] = `"2009-05-19T14:39:22-06:00 = Tue May 19 2009 20:39:22 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-05-19T14:39Z" successfully 1`] = `"2009-05-19T14:39Z = Tue May 19 2009 14:39:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009-12T12:34" successfully 1`] = `"2009-12T12:34 = Tue Dec 01 2009 12:34:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "200905" successfully 1`] = `"200905 = Fri May 01 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "200912T1234" successfully 1`] = `"200912T1234 = Tue Dec 01 2009 12:34:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "2009001" successfully 1`] = `"2009001 = Fri Jan 02 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20070406 0000" successfully 1`] = `"20070406 0000 = Fri Apr 06 2007 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519 0000" successfully 1`] = `"20090519 0000 = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519 1431" successfully 1`] = `"20090519 1431 = Tue May 19 2009 14:31:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519 1439Z" successfully 1`] = `"20090519 1439Z = Tue May 19 2009 14:39:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519 143922" successfully 1`] = `"20090519 143922 = Tue May 19 2009 14:39:22 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519 143922+0600" successfully 1`] = `"20090519 143922+0600 = Tue May 19 2009 08:39:22 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519 143922-06:00" successfully 1`] = `"20090519 143922-06:00 = Tue May 19 2009 20:39:22 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519" successfully 1`] = `"20090519 = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;

exports[`Parsing without a binding Parses iso-dates  parse "20090519" successfully 2`] = `"20090519 = Tue May 19 2009 00:00:00 GMT+0000 (Greenwich Mean Time)"`;
